src/background.c

```diff
@@ -0,0 +1,348 @@
+      return;
+    }
+  permissions[1] = NULL;
+
+  if (!xdp_impl_permission_store_call_set_permission_sync (get_permission_store (),
+                                                           PERMISSION_TABLE,
+                                                           TRUE,
+                                                           PERMISSION_ID,
+                                                           app_id,
+                                                           (const char * const*)permissions,
+                                                           NULL,
+                                                           &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_warning ("Error updating permission store: %s", error->message);
+    }
+}
+
+static void
+handle_request_background_in_thread_func (GTask *task,
+                                          gpointer source_object,
+                                          gpointer task_data,
+                                          GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  GVariant *options;
+  const char *app_id;
+  Permission permission;
+  gboolean allowed;
+  const char *reason = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  options = (GVariant *)g_object_get_data (G_OBJECT (request), "options");
+  g_variant_lookup (options, "reason", "s", &reason);
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission (app_id);
+
+  g_debug ("Handle RequestBackground for %s\n", app_id);
+
+  if (permission == ASK || permission == UNSET)
+    {
+      GVariantBuilder opt_builder;
+      g_autofree char *title = NULL;
+      g_autofree char *subtitle = NULL;
+      g_autofree char *body = NULL;
+      guint32 response = 2;
+      g_autoptr(GVariant) results = NULL;
+      g_autoptr(GError) error = NULL;
+      g_autoptr(GAppInfo) info = NULL;
+
+      if (app_id[0] != 0)
+        {
+          g_autofree char *desktop_id;
+          desktop_id = g_strconcat (app_id, ".desktop", NULL);
+          info = (GAppInfo*)g_desktop_app_info_new (desktop_id);
+        }
+
+      title = g_strdup_printf (_("Allow %s to run in the background?"), info ? g_app_info_get_display_name (info) : app_id);
+      if (reason)
+        subtitle = g_strdup (reason);
+      else
+        subtitle = g_strdup_printf (_("%s requests to run in the background."), info ? g_app_info_get_display_name (info) : app_id);
+      body = g_strdup (_("The ‘run in background’ permission can be changed at any time from the application settings."));
+
+      g_debug ("Calling backend for background access for: %s", app_id);
+
+      g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+      g_variant_builder_add (&opt_builder, "{sv}", "deny_label", g_variant_new_string (_("Don't allow")));
+      g_variant_builder_add (&opt_builder, "{sv}", "grant_label", g_variant_new_string (_("Allow")));
```

Comment:
Do we never want the notification to be able to reply "always ask"? (or "allow once")

Reply:
The portal we use here is just asking a deny/grant question, atm. And the designers don't like the 'always ask' much. They prefer to have permissions that time out.